### ISSUE TITLE
[Camt.054] Returned dates is the current one instead of NULL if not part of the XML

### ISSUE THREAD

**Issue by sdespont** [genkgo/camt#61]
In case of some optional dates are missing in the XML file, the date returned by the library is **the current one** instead of `NULL`.

For example, in the XML validation schema point of view, bookingDate and valueDate are optionals. But Theses dates are mandatory in case of CAMT.054 file. If one of these dates are not present, the returned date is the current one, which is obviously false.

I would suggest to test if the date is empty in the date decoder and return NULL if it is the case instead of the current date : 
```
    /**
     * @param $date
     * @return \DateTimeImmutable|null
     */
    public function decode($date)
    {
        if(empty($date)) {
            return null;
        }

        if ($this->format === null) {
            $result = new \DateTimeImmutable($date);
        } else {
            $result = \DateTimeImmutable::createFromFormat($this->format, $date);
        }

        if ($result === false) {
            throw new \InvalidArgumentException("Cannot decode date {$date}");
        }

        return $result;
    }
```

But in that case, the signature of `Entry `constructor must be modified to remove the DateTimeImmutable expected class for $bookingDate and $valueDate parameters because they can be also `NULL`.

What do you think about this suggestion?

**Comment by frederikbosch** [comment]
Agree.

**Comment by sdespont** [comment]
Ok, I will create a pull request for this.

**Comment by sdespont** [comment]
But, it will break compatibility with all objects constructors using `DateTimeImmutable` as parameter. Perhaps the best solution is to create a new method `decodeWihoutDefaultValue` in the `DateDecoderInterface `and call it for `bookingDate `and `valueDate `in Record.php?

https://github.com/genkgo/camt/blob/670d7e28ec9efb68951f7e1d6f5219f02c3748ba/src/Decoder/Record.php#L93-L100

**Comment by frederikbosch** [comment]
This makes CAMT parsing so hard. Actually, this requires to go for a solution like this.

```php
$entry = new DTO\Entry( 
  $record, 
  $index, 
  new Money($amount, new Currency($currency))
); 
$entry->setBookingDate($this->dateDecoder->decode($bookingDate));
$entry->setValueDate($this->dateDecoder->decode($bookingDate));
$entry->setAdditionalInfo($additionalInfo );
```

**Comment by sdespont** [comment]
Sorry, I don't get your your point. 

It is not just about the constructor of `DTO\Entry`. The fact is if `$this->dateDecoder->decode` can return in the future `DateTimeImmutable `or `NULL`, all methods using dates in theirs definitions could throw an exception because the given date is `NULL `instead of `DateTimeImmutable `.

To avoid BC, I would using like this : 

```
    /**
     * @param string $date
     * @param boolean $useNowIfEmpty
     * @return \DateTimeImmutable|null
     */
    public function decode($date, $useNowIfEmpty = true)
    {
        // Do not use current date if the date is empty
        if(empty($date)
                && $useNowIfEmpty === false) {
            return null;
        }

        if ($this->format === null) {
            $result = new \DateTimeImmutable($date);
        } else {
            $result = \DateTimeImmutable::createFromFormat($this->format, $date);
        }

        if ($result === false) {
            throw new \InvalidArgumentException("Cannot decode date {$date}");
        }

        return $result;
    }
```

Then, use this new parameter in the calls where current date **must not** be returned in case of missing date in the XML file : 

```
$entry = new DTO\Entry( 
                 $record, 
                 $index, 
                 new Money($amount, new Currency($currency)), 
                 $this->dateDecoder->decode($bookingDate, false), 
                 $this->dateDecoder->decode($valueDate, false), 
                 $additionalInfo 
             ); 
```

In that way, there is no BC. My opinion is the `date ` decode must always returns `NULL `in case of missing date, but I understand that it would BC for some users and it is ncessary to deal with that.

What is you opinion?

**Comment by frederikbosch** [comment]
@sdespont Sorry, my explanation was too short.

```php
$entry = new DTO\Entry( 
  $record, 
  $index, 
  new Money($amount, new Currency($currency))
); 
if ($bookingDate) {
  $entry->setBookingDate($this->dateDecoder->decode($bookingDate));
}
if ($valueDate) {
  $entry->setValueDate($this->dateDecoder->decode($valueDate));
}
$entry->setAdditionalInfo($additionalInfo );
```

So, we should not call the date decoder when there is no date. That leaves the date decoder simple, and solves the solution where to problem emerges: the XML.

**Comment by sdespont** [comment]
Ok, that makes sense. I agree that letting the decoder simple is a better solution. 

I will create a pull request soon.

**Comment by PowerKiKi** [comment]
Should be fixed by #62
